### PR TITLE
ARROW-1765: [Doc] Use dependencies from conda in C++ docker build

### PR DESCRIPTION
**Pull request by xhochy**

### CODE REVIEW

File: dev/gen_apidocs/Dockerfile [hunk 1]
```diff
@@ -79,5 +70,9 @@ RUN /home/ubuntu/miniconda/bin/conda create -y -q -n pyarrow-dev \
         jemalloc \
         lz4-c \
         zstd \
+        doxygen \
+        maven \
         -c conda-forge
+ADD . /apache-arrow
```
Comment (xhochy) [inline dev/gen_apidocs/Dockerfile]: Moving this here improves Docker's caching possibilities.

**Comment by cpcloud** [comment]
@kou It looks like there's a failing static assertion in Glib: https://travis-ci.org/apache/arrow/jobs/296512183
This is also happening in #1267 here: https://travis-ci.org/apache/arrow/jobs/296576140

Can you take a look?

**Comment by kou** [comment]
It seems that 32bit binary (universal binary) is trying to build implicitly.
Please try `ARCHFLAGS="-arch x86_64" make` on macOS for building Arrow GLib.

**Review by xhochy — COMMENTED** [review]
+1

**Pull request closed without merging**